### PR TITLE
Fixes discard_memory compilation failure for pre-Volta

### PR DESCRIPTION
**Pull request by elstehle**
## Description

Closes issue https://github.com/NVIDIA/cccl/issues/624 and unblocks PR https://github.com/NVIDIA/cccl/pull/619

Including `<cuda/annotated_ptr>` includes `<cuda/barrier>`, which in turn includs `"std/barrier"`. `"std/barrier"` in turn has the following guard in place:
```
#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 700
#  error "CUDA synchronization primitives are only supported for sm_70 and up."
#endif
```
This prevents users from using functionality like `cuda::discard_memory` in `<cuda/annotated_ptr>` that does not rely on `barrier`.

This PR:
- moves `cuda::discard_memory` to its own header
- fixes aligned pointer computation in `cuda::discard_memory` and the loop for discarding memory
- fixes initialization order in `__access_property` that the compiler complained about

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: libcudacxx/include/cuda/discard_memory [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+ * NVIDIA, 2788 San Tomas Expressway, Santa Clara, CA 95051.
+ *
+ * 15. ENTIRE AGREEMENT. This license is the final, complete and exclusive agreement between the parties relating to the
+ * subject matter of this license and supersedes all prior or contemporaneous understandings and agreements relating to
+ * this subject matter, whether oral or written. If any court of competent jurisdiction determines that any provision of
+ * this license is illegal, invalid or unenforceable, the remaining provisions will remain in full force and effect.
+ * This license may only be modified in a writing signed by an authorized representative of each party.
+ *
+ * (v. August 20, 2021)
+ */
```
Comment (miscco) [inline libcudacxx/include/cuda/discard_memory:130]: This should use our shortened license similar to other headers

File: libcudacxx/include/cuda/discard_memory [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+#ifndef _CUDA_DISCARD_MEMORY
+#define _CUDA_DISCARD_MEMORY
+
+#include <cuda/std/cstdint>
+#include "std/detail/__config"
```
Comment (miscco) [inline libcudacxx/include/cuda/discard_memory:136]: This should either be
```suggestion
#include "std/cstdint"
#include "std/detail/__config"
```
or
```suggestion
#include <cuda/std/cstdint>
#include <cuda/std/detail/__config>
```

**Review by miscco — APPROVED** [review]

File: libcudacxx/include/cuda/discard_memory (hunk 2 above)
Comment (elstehle) [reply libcudacxx/include/cuda/discard_memory]: Good catch. Fixed the includes and went with your second suggestion 👍

File: libcudacxx/include/cuda/discard_memory (hunk 1 above)
Comment (elstehle) [reply libcudacxx/include/cuda/discard_memory]: Thanks, I switched to the shortened license 👍

**Pull request merged**